### ISSUE TITLE
${prompt.text} and ${prompt.secret} double prompting

### ISSUE THREAD

**Issue by djschny** [elastic/elasticsearch#11564]
With the following configuration
```
node.name: ${prompt.text}
```
Elasticsearch 1.6.0 prompts you twice. Once for "node.name" and "name". Ultimately it uses the second one for the value of the configuration item:
```
djschny:elasticsearch-1.6.0 djschny$ ../startElastic.sh 
Enter value for [node.name]: foo
Enter value for [name]: bar
[2015-06-09 16:10:03,405][INFO ][node                     ] [bar] version[1.6.0], pid[4836], build[cdd3ac4/2015-06-09T13:36:34Z]
[2015-06-09 16:10:03,405][INFO ][node                     ] [bar] initializing ...
```

**Comment by clintongormley** [comment]
@jaymode please could you take a look

**Comment by joshuar** [comment]
Looks to be still a problem on 2.1.  Regression here?  @GlenRSmith and I can confirm this occurs with the 2.1 release archive.

**Comment by joshuar** [comment]
It seems to be independent of which config item you want to prompt for:

```
elasticsearch-2.1.0  bin/elasticsearch
Enter value for [cluster.name]: foo
Enter value for [cluster.name]: bar
[2015-12-18 10:48:50,219][INFO ][node                     ] [Wendell Vaughn] version[2.1.0], pid[21231], build[72cd1f1/2015-11-18T22:40:03Z]
[2015-12-18 10:48:50,220][INFO ][node                     ] [Wendell Vaughn] initializing ...
[2015-12-18 10:48:50,280][INFO ][plugins                  ] [Wendell Vaughn] loaded [], sites []
[2015-12-18 10:48:50,304][INFO ][env                      ] [Wendell Vaughn] using [1] data paths, mounts [[/ (/dev/mapper/fedora_josh--xps13-root)]], net usable_space [79.1gb], net total_space [233.9gb], spins? [no], types [ext4]
[2015-12-18 10:48:51,761][INFO ][node                     ] [Wendell Vaughn] initialized
[2015-12-18 10:48:51,762][INFO ][node                     ] [Wendell Vaughn] starting ...
[2015-12-18 10:48:51,902][INFO ][transport                ] [Wendell Vaughn] publish_address {127.0.0.1:9300}, bound_addresses {127.0.0.1:9300}, {[::1]:9300}
[2015-12-18 10:48:51,912][INFO ][discovery                ] [Wendell Vaughn] bar/azOFxYXMQ6muCpGOpqvxSw
```

**Comment by GlenRSmith** [comment]
Just confirmed on 2.1.1.

**Comment by rjernst** [comment]
This is different than the original issue. I think what is happening is we first initialize the settings/environment in bootstrap so that we can eg init logging, but then when bootstrap creates the node, it passes in a fresh Settings, and the Node constructor again initializes settings/environment.

**Comment by clintongormley** [comment]
@jaymode could you take a look at this please?

**Comment by jaymode** [comment]
As @rjernst said, this is a different issue that the original. `BootstrapCLIParser` was added which extends CLITool. CLITool prepares the settings and environment which includes prompting since CLITools are usually run outside of the bootstrap process. This causes the first prompt that is ignored. The second prompt that is used, comes from `Bootstrap`, which is passed to the node.

Part of the issue is that the `BootstrapCLIParser` sets properties that will change the value of settings. I think we can solve this a few different ways:

1. Pass in empty settings/null environment for this CLITool. If we try to create a valid environment here then we have to prepare the settings to ensure we parse the paths from the settings for our directories
2. Do not use the `CLITool` infrastructure
3. Prepare the environment in bootstrap, pass to `BootstrapCLIParser`. Re-prepare the settings/environment, passing in the already prepared settings.

@spinscale @rjernst any thoughts?

**Comment by rjernst** [comment]
I would say preparing the environment once would be the best option, it will just take some refactoring to pass it through. Running prepare already creates Environment twice (the first time so it can try and load the config file, which might have other paths like plugin path or data path). Really I think we should simply not allow paths to be configured in elasticsearch.yml, instead it should only be through sysprops.  It might be something that could simplify the settings/env prep, to make this a little easier.

**Comment by jasontedor** [comment]
I have a fix for this that will come as part of #16579.

**Comment by jasontedor** [comment]
Closed by #17024.
